regression with portable-atomics in `1.20.0`

#260 propagates the `critical-section` feature to portable-atomics, which conflicts with `unsafe-assume-single-core`.

> * propagate `critical-section` feature selection into `portable-atomic`

This breaks our use case. `portable-atomic` treats `critical-section` and `unsafe-assume-single-core` as mutually exclusive. We depend on esp-rs, which for some platforms (e.g., esp32c2, c3) explicitly enables the latter.

It is my understanding that `libraries` should not set either `critical-section` or `unsafe-assume-single-core`, but leave that to the end user.

#260 states this:

```
ORIGINAL RATIONALE RE: propagate critical-section feature selection into portable-atomic

Normally just using critical-section should be good enough for targets like super-outdated thumbv6m-none-eabi target. But in https://github.com/rustls/rustls/pull/2088 I would like to preserve existing use of alloc option, in which case I had to add explicit portable-atomic dependency with its critical-section feature enabled.

IMO it would be nice if I didn't have to add explicit sub-depencencies with this option. This is my proposal to enable critical-section for portable-atomic if portable-atomic is wanted.

Please let me know if you have any other idea or perspective concerning this.
```

What does `critical-section` on `portable-atomics` actually enable that only `critical-section` feature enable? Maybe this can be solved in `portable-atomic` with a backend-agnostic feature like `require-cas`?

Here is my understanding on the implementation before https://github.com/matklad/once_cell/pull/260 (i.e., 1.19.0):

First, there are two no-std compatible thread-safe implementations and each depends on different things:

- `once_cell::sync`: using `critical-section` and atomic load/store (provided by `portable-atomic`)
  - `critical-section` is mandatory for this implementation.
  - `portable-atomic` is *not* mandatory for this implementation because atomic load/store is available in `core` except for MSP430, PSX, and BPF targets.
  - If `portable-atomic` is available as a dependency, it would be better to use `portable-atomic`, because it can support the above targets that atomic load/store is not available in `core`.
- `once_cell::race`: using atomic CAS (provided by `portable-atomic` if `critical-section` feature is enabled)
  - `critical-section` is *not* used for this implementation.
  - `portable-atomic` is *not* mandatory for this implementation because atomic CAS is available in `core` except for thumbv6m, pre-v6 Arm, RISC-V without A extension, Xtensa, AVR, MSP430, MIPS PSX, and BPF targets.
  - If `portable-atomic` is available as a dependency, it would be better to use `portable-atomic`, because it can support the above targets that atomic CAS is not available in `core`.

And I think there are two issues:

- Using `portable-atomic` in `once_cell::race` (to make it available on targets that don't support atomic CAS) requires `critical-section` feature but it is confusing because `critical-section` is not used at all in `once_cell::race`.
- `once_cell::race` requires atomic CAS, but `portable-atomic` is not always able to provide it, and error messages when `require-cas` feature of `portable-atomic` is not enabled are confusing.

For the first issue, it is better to allow using `portable-atomic` without `critical-section` feature.
I think there are two ways:
- New Cargo feature to use `portable-atomic`'s atomic types instead of `core`'s:
  ```toml
  [features]
  race = ["portable-atomic?/require-cas"]
  critical-section = ["dep:critical-section", "portable-atomic"]

  [dependencies]
  portable-atomic = { version = "1.3", optional = true, default-features = false }
  ```
  ```rs
  #[cfg(not(feature = "portable-atomic"))]
  use core::sync::atomic::AtomicPtr;
  #[cfg(feature = "portable-atomic")]
  use portable_atomic::AtomicPtr;
  ```
- Target-specific dependency to use `portable-atomic`'s atomic types on `cfg(not(target_has_atomic = "ptr"))` targets (no new Cargo feature):
  ```toml
  [features]
  race = ["dep:portable-atomic", "portable-atomic/require-cas"]
  critical-section = ["dep:critical-section", "dep:portable-atomic"]

  [target.'cfg(not(target_has_atomic = "ptr"))'.dependencies]
  portable-atomic = { version = "1.3", optional = true, default-features = false }
  ```
  ```rs
  #[cfg(target_has_atomic = "ptr")]
  use core::sync::atomic::AtomicPtr;
  #[cfg(not(target_has_atomic = "ptr"))]
  use portable_atomic::AtomicPtr;
  ```

(Note that `critical-section` feature needs to continue to have `portable-atomic` enabled for compatibility with previous releases.)

For the second issue, adding `require-cas` feature as above is better, but I have found that it is possible to display decent error messages without `require-cas` feature using `#[diagnostic::on_unimplemented]` stabilized in Rust 1.78: https://github.com/taiki-e/portable-atomic/issues/180

- Before:

  ```
  error[E0599]: no method named `compare_exchange` found for struct `portable_atomic::AtomicUsize` in the current scope
    --> src/race.rs:60:24
     |
  60 |             self.inner.compare_exchange(0, value.get(), Ordering::AcqRel, Ordering::Acquire);
     |                        ^^^^^^^^^^^^^^^^ method not found in `AtomicUsize`
  ```
- After:
  
  ```
  error[E0277]: `compare_exchange` requires atomic CAS but not available on this target by default
      --> src/race.rs:60:24
       |
  60   |             self.inner.compare_exchange(0, value.get(), Ordering::AcqRel, Ordering::Acquire);
       |                        ^^^^^^^^^^^^^^^^ `compare_exchange` requires atomic CAS but not available on this target by default
       |
       = help: the trait `HasCompareExchange` is not implemented for `&portable_atomic::AtomicUsize`
       = note: consider enabling one of the `unsafe-assume-single-core` or `critical-section` Cargo features
       = note: see <https://docs.rs/portable-atomic/latest/portable_atomic/#optional-features> for more.
  note: required by a bound in `portable_atomic::AtomicUsize::compare_exchange`
      --> /Users/taiki/projects/sources/taiki-e/portable-atomic/src/lib.rs:4064:5
       |
  4064 |     atomic_int!(AtomicUsize, usize, 4);
       |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       |     |
       |     required by a bound in this associated function
       |     required by this bound in `AtomicUsize::compare_exchange`
       = note: this error originates in the macro `atomic_int` (in Nightly builds, run with -Z macro-backtrace for more info)
  ```
---

@brodycj

> maybe rustls should offer the "unsafe" single core option as well?

I don't think a feature for `unsafe-assume-single-core` in rustls is needed. [It has various associated features that control its behavior (annoying to forward all) and it can also be set globally via cfg](https://github.com/taiki-e/portable-atomic/blob/HEAD/src/imp/interrupt/README.md).

--- a/Cargo.lock.msrv
+++ b/Cargo.lock.msrv
@@ -69,9 +69,6 @@ name = "portable-atomic"
 version = "1.7.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "da544ee218f0d287a911e9c99a39a8c9bc8fcad3cb8db5959940044ecfc67265"
-dependencies = [
- "critical-section",
-]
 
 [[package]]
 name = "redox_syscall"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -21,7 +21,7 @@ members = ["xtask"]
 
 [dependencies]
 parking_lot_core = { version = "0.9.10", optional = true, default-features = false }
-portable-atomic = { version = "1.7", optional = true }
+portable-atomic = { version = "1.7", optional = true, default-features = false }
 critical-section = { version = "1.1.3", optional = true }
 
 [dev-dependencies]
@@ -38,17 +38,23 @@ std = ["alloc"]
 alloc = ["race"]
 
 # Enables `once_cell::race` module.
-race = []
+race = ["portable-atomic?/require-cas"]
 
 # Uses parking_lot to implement once_cell::sync::OnceCell.
 # This makes no speed difference, but makes each OnceCell<T>
 # up to 16 bytes smaller, depending on the size of the T.
 parking_lot = ["dep:parking_lot_core"]
 
-# Uses `critical-section` to implement `sync` and `race` modules. in
+# Uses `portable-atomic` to implement `race` module. in
+# `#![no_std]` mode. Please read `portable-atomic` docs carefully
+# before enabling this feature.
+portable-atomic = ["dep:portable-atomic"]
+
+# Uses `critical-section` to implement `sync` module. in
 # `#![no_std]` mode. Please read `critical-section` docs carefully
 # before enabling this feature.
-critical-section = ["dep:critical-section", "portable-atomic/critical-section"]
+# `portable-atomic` feature is enabled for backwards compatibility.
+critical-section = ["dep:critical-section", "portable-atomic"]
 
 # Enables semver-exempt APIs of this crate.
 # At the moment, this feature is unused.

--- a/src/race.rs
+++ b/src/race.rs
@@ -19,9 +19,9 @@
 //! `Acquire` and `Release` have very little performance overhead on most
 //! architectures versus `Relaxed`.
 
-#[cfg(not(feature = "critical-section"))]
+#[cfg(not(feature = "portable-atomic"))]
 use core::sync::atomic;
-#[cfg(feature = "critical-section")]
+#[cfg(feature = "portable-atomic")]
 use portable_atomic as atomic;
 
 use atomic::{AtomicPtr, AtomicUsize, Ordering};

